[DevOps] Add devops/scripts/sycl-bisect.bash

dwoodwor-intel opened the pull request:
This is a utility for running git-bisect on sycl-like branches that may
contain upstream LLVM commits in the bisect range. Its main feature is
automatically merging these commits into a sycl-based branch before
testing them, and supports conveniently testing commits by specifying a
LIT test to run or a custom bash command.

aelovikov-intel commented on buildbot/sycl-bisect.bash:
```diff
@@ -0,0 +1,195 @@
+  # Try the bad rev first.
+  check_sycl_descendant "${REQUIRED_ANCESTORS[1]}" && return
+
+  # Try "sycl" and "sycl-web", both locally and in all the remotes.
```
nit: should we limit that to official remotes only? If, for some reason, somebody uses ordinary development workspace there might be extra remotes that wouldn't make sense (e.g. my github remote has *very* old sycl/sycl-web).

dwoodwor-intel replied:
This is just looking for _any_ commit that is descended from REQUIRED_ANCESTORS, so extra remotes shouldn't cause any issues besides slowing down this check. This is part of an elaborate workaround to try to approximate a version of git-merge-base that finds common descendants instead of ancestors, which is usually not a well-defined question in git because you can have commits merged differently in different branches. In our usage that's very unlikely, though, so any branch we can find that's descended from the required commits is good enough for figuring out where to merge. We still probably don't want to test _every_ remote branch because there can be a lot of them and it would take forever to find out if none of them have all of the commits, but just checking the "sycl" and "sycl-web" branches of each remote seems like a good compromise to still be able to catch all of the most likely remote branches.

aelovikov-intel reviewed (commented):
This looks great. Honestly, it's more beneficial for me to read this than any my feedback to you!

LGTM, but no approval in case someone else has something to say...

cperkinsintel commented on buildbot/sycl-bisect.bash:
```diff
@@ -0,0 +1,195 @@
+#!/bin/bash
+
+# This is a script for running git-bisect on sycl, sycl-web, and sycl pulldown
```
These docs are good, but could you include an example of running it?
When bisecting "normally" I usually just do `git bisect start` `git bisect good xxxx` `git bisect bad xxxx` `git bisect run my_test_script.sh` . But its unclear to me exactly how to use this one, maybe because I'm not appreciating the context of its use fully, having not worked with the llvm pulldown difficulties very much.

dwoodwor-intel replied:
I've added a usage example to the help text. I'm not sure if you're aware of a similar tool we have for our internal compiler repository called "Xmain-aware Git Bisect", but the way this tool is run is inspired by that. It's very similar to the normal git-bisect procedure, except that you specify the bad commit, good commit(s), and what to test all in one step instead of in multiple stages.

You can also use test-commit-sycl-bisect.bash standalone with the normal git-bisect flow if you use the environment variables as described at the top of that file, but this is a more advanced usage and you miss out on some of the up-front setup/validation that sycl-bisect.bash does. Also, if test-commit-sycl-bisect.bash changes/doesn't exist somewhere in your bisection range this can cause issues (which sycl-bisect.bash works around by copying test-commit-sycl-bisect.bash to a temporary file before starting the bisection).

bader reviewed (commented):
I suggest we put these to `devops/scripts`. We don't use BuildBot CI anymore, so I expect devops team to delete this directory and move scripts from there to other directories.

dwoodwor-intel commented:
I've moved these into `devops/scripts`, but they still have references to `buildbot/configure.py` and `buildbot/compile.py` that should be updated when those scripts are also moved (probably with a fallback so that sycl-bisect can still be used on commits with the old script locations). How soon will that be?

bader commented:
> I've moved these into `devops/scripts`, but they still have references to `buildbot/configure.py` and `buildbot/compile.py` that should be updated when those scripts are also moved (probably with a fallback so that sycl-bisect can still be used on commits with the old script locations). How soon will that be?

I think this is the question for the Devops team (@intel/dpcpp-devops-reviewers?). 

I suggest we move these scripts in two steps to ensure smooth transition:
1. Create `devops` versions of these scripts in new location, update documentation with instructions to use new scripts and add a deprecation message to buildbot script.
2. Remove buildbot directory.

I think there might be users of these scripts how need some time on updating the way we configure DPC++ compiler build.

One more note: I'm don't see much value in `check.py` and `compile.py`, which are simple wrappers for `cmake`. `configure.py` can also be replaced with cmake presets or cmake cache for old cmake versions (e.g. see clang project cmake cache files as example).

bader commented on devops/scripts/sycl-bisect.bash:
```diff
@@ -0,0 +1,205 @@
+# Make sure the build directory is set up; if not, tell the user to run
+# the configure step.
+if [[ ! -f build/CMakeCache.txt ]]; then
+  echo "The build directory doesn't seem to be configured yet." >&2
+  echo "Please run the configure step to set up the build." >&2
+  exit 1
```
To make this message more you useful, you can consider adding a reference to the GetStartedGuide, where we have up-to-date valid configuration instructions.

E.g. https://intel.github.io/llvm-docs/GetStartedGuide.html#build-dpc-toolchain or https://github.com/intel/llvm/blob/sycl/sycl/doc/GetStartedGuide.md#build-dpc-toolchain

dwoodwor-intel replied:
I've updated this message to reference the documentation
